### PR TITLE
Improved performance for TrimRows, HiddenRows and HiddenColumns plugins for big datasets with lots trimmed/hidden indexes

### PR DESCRIPTION
**Pull request by wszymanski**
### Context
While investigating an example done by @bamminger I've discovered that it seems that there are at least two bottlenecks:

- First one related to the `TrimRows` plugin.
- Second one related to the `AutoColumnSize` plugin (should be done within the [newly created issue](https://github.com/handsontable/handsontable/issues/7235))

Screenshot 2020-08-26 at 16 08 48

Within this PR I've focused on improving the work of `TrimRows` plugin (and consistently for `HiddenColumns` and `HiddenRows` plugins) for the example with quite a big dataset, mentioned by the user. 

The current code tries to filter the list of already trimmed indexes from untrimmed indexes. 

https://github.com/handsontable/handsontable/blob/0ec29f9cab9e0d9df1af99287c1211e87779737a/src/plugins/trimRows/trimRows.js#L172
https://github.com/handsontable/handsontable/blob/0ec29f9cab9e0d9df1af99287c1211e87779737a/src/plugins/trimRows/trimRows.js#L177

As an effect, there can be performed `NxM` searches, where `N` is the number of already trimmed indexes, while `M` is the number of untrimmed indexes. These operations have very high computational complexity. An example provided by the user shows the worst-case scenario when there are performed almost `100k^2` array searches.

I've changed a way how we create a list of trimmed indexes right after the `untrim` action. As an effect, I've reduced the time of code execution for `untrim` on my computer for some cases.

I've tested how browsers perform mainly on my computer - MacBookPro (15", Mid 2015, i7 2,2 GHz), macOS Catalina. Just tests for IE 11 have been performed using Browserstack. I've skipped some tests when the difference was very low/very big.

### Average results for tested browsers (compared to the old codebase)

| dataset (number of rows) | Google Chrome 84        | Firefox 79  | Safari 13.1.1      | IE 11      |
|--------------------------|-------------------------|-------------|--------------------|------------|
| 100k, 10 trimmed (0,01%) | **↓ 7%** ❗             | ↑ 30%       | ↑ 13%              | **↓ 1%** ❗ |
| 100k, 100 trimmed (0,1%) | **↓ 3%** ❗             | ↑ 24%       | **↓ 7%** ❗        | **↓ 2%** ❗ |
| 100k, 1k trimmed (1%)    | **↓ 5%** ❗             | ↑ 34%       | ↓ 0,5%             | ↑ 1%       |
| 100k, 10k trimmed (10%)  | ↑ 89%                   | ↑ 51%       | ↑ 35%              | ↑ 475%     |
| 100k, 50k trimmed (50%)  | ↑ 2 424%                | ↑ 1 362%    | ↑ 710%             | -          |

| dataset (number of rows)  | Google Chrome 84 | Firefox 79 | Safari 13.1.1 | IE 11     |
|---------------------------|-----------------|------------|---------------|--------------|
| 10k, 1 trimmed (0,01%)    | ↑ 58%           | -          | -             | ↑ 8%         |
| 10k, 10 trimmed (0,1%)    | ↑ 61%           | -          | -             | **↓ 1%** ❗  |
| 10k, 100 trimmed (1%)     | ↑ 27%           | -          | -             | **↓ 6%** ❗  |
| 10k, 1k trimmed (10%)     | ↑ 99%           | -          | -             | ↑ 38%        |
| 10k, 5k trimmed (50%)     | ↑ 235%          | ↑ 262%     | ↑ 150%        | -            |

### Conclusions

As we can see in the above table there is a huge difference in the execution time for tables with a lot of trimmed rows. However, some browsers don't perform very well when there are less trimmed rows.

I will go through them:

- For Google Chrome with `100 000` rows and `10`, `100` and `1 000` trimmed rows there was a difference for throttled `6` times CPU by Chrome DevTools, respectively: ~`30ms`/`6ms`/`22ms` which may increase for mobile devices, but seems to be not such big numbers.
- For Safari with `100 000` rows and `100` trimmed there was a difference of ~`11ms` which seems to be not such a big number.
- IE11 was lagged a lot, some actions took even some seconds on the hosted machine, I would not consider it.

I would choose the proposed idea - it's quite stable and gives a big influence on large datasets in some specific cases. For the case mentioned by the user, code execution time has been decreased from a few seconds to just a hundred of milliseconds.

### How has this been tested?

Tested code execution time by `console.time`. Real change can be perceived organoleptically when the `AutoColumnSize` plugin is disabled (it's enabled by default).

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #6420

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: src/plugins/hiddenColumns/hiddenColumns.js [hunk 1]
```diff
@@ -157,7 +157,7 @@ class HiddenColumns extends BasePlugin {
     let destinationHideConfig = currentHideConfig;
 
     if (isConfigValid) {
-      destinationHideConfig = currentHideConfig.filter(column => columns.includes(column) === false);
+      destinationHideConfig = Array.from(new Set(this.#hiddenColumnsMap.getNotHiddenIndexes().concat(columns)));
```
Comment (wojciechczerniak) [inline src/plugins/hiddenColumns/hiddenColumns.js:160]: The `Array.from(new Set(` pattern is repeating three times in this PR.

Shouldn't we move common code or use one of the array helpers? https://github.com/handsontable/handsontable/blob/560eb12677ec1b6b02d86b1f7d9715410f69d903/src/helpers/array.js#L246-L317


and optimize it there with `Array.from(new Set(` pattern so all use cases across HoT are optimized as well?

----

i.e.

```suggestion
      destinationHideConfig = arrayUnique(this.#hiddenColumnsMap.getNotHiddenIndexes().concat(columns));
```

or without `concat`:

```suggestion
      destinationHideConfig = getUnionOfArrays(this.#hiddenColumnsMap.getNotHiddenIndexes(), columns);
```

File: src/plugins/hiddenRows/hiddenRows.js [hunk 2]
```diff
@@ -158,7 +158,7 @@ class HiddenRows extends BasePlugin {
     let destinationHideConfig = currentHideConfig;
 
     if (isConfigValid) {
-      destinationHideConfig = currentHideConfig.filter(row => rows.includes(row) === false);
+      destinationHideConfig = Array.from(new Set(this.#hiddenRowsMap.getNotHiddenIndexes().concat(rows)));
```
Comment (wojciechczerniak) [inline src/plugins/hiddenRows/hiddenRows.js:161]: duplicated code, see previous comment

File: src/plugins/trimRows/trimRows.js [hunk 3]
```diff
@@ -174,7 +174,7 @@ class TrimRows extends BasePlugin {
     let destinationTrimConfig = currentTrimConfig;
 
     if (isValidConfig) {
-      destinationTrimConfig = currentTrimConfig.filter(trimmedRow => rows.includes(trimmedRow) === false);
+      destinationTrimConfig = Array.from(new Set(this.trimmedRowsMap.getNotTrimmedIndexes().concat(rows)));
```
Comment (wojciechczerniak) [inline src/plugins/trimRows/trimRows.js:177]: duplicated code, see previous comment

File: src/translations/maps/indexMap.js [hunk 4]
```diff
@@ -121,6 +122,23 @@ class IndexMap {
...
+   * @returns {Array}
+   */
+  getIndexesByValue(value) {
+    return arrayReduce(this.getValues(), (indexesList, valueInMap, index) => {
+      if (valueInMap === value) {
+        indexesList.push(index);
+      }
+
+      return indexesList;
+    }, []);
```
Comment (wojciechczerniak) [inline src/translations/maps/indexMap.js:139]: `arrayFilter` will not be more convenient helper for this use case? https://github.com/handsontable/handsontable/blob/560eb12677ec1b6b02d86b1f7d9715410f69d903/src/helpers/array.js#L102

```suggestion
    return arrayFilter(this.getValues(), (valueInMap) => {
      return valueInMap === value;
    });
```

Comment (wszymanski) [reply src/translations/maps/indexMap.js]: Nope. We do not filter booleans, we try to get indexes.

File: src/plugins/hiddenColumns/hiddenColumns.js (hunk 1 above)
Comment (wszymanski) [reply src/plugins/hiddenColumns/hiddenColumns.js:160]: We discussed about that some times ago. Maybe some extra issue should be done for that? It should be tested properly.

PS. My code has been changed. It won't be used anymore. However, a hint about code duplication may be still accurate. I will think whether some extra methods/helpers should be added.

File: src/translations/maps/indexMap.js (hunk 4 above)
Comment (wojciechczerniak) [reply src/translations/maps/indexMap.js]: Right!

File: src/plugins/hiddenColumns/hiddenColumns.js (hunk 1 above)
Comment (wszymanski) [reply src/plugins/hiddenColumns/hiddenColumns.js]: I'm not sure if parts of the current code should be moved to one place. I have no idea how I can extend `TrimmingMap` and `HidingMap` - whether some generic methods for getting **copy** of the state after `untrim`/`unhide` should be provided 🤔

File: src/plugins/hiddenColumns/hiddenColumns.js [hunk 5]
```diff
@@ -153,33 +153,45 @@ class HiddenColumns extends BasePlugin {
...
+
+      // Preparing new values for hiding map.
+      arrayEach(physicalColumns, (physicalColumn) => {
+        hidingMapValues[physicalColumn] = false;
+      });
+
+      // Preparing new hiding config.
+      destinationHideConfig = arrayReduce(hidingMapValues, (hiddenIndexes, isHidden, physicalIndex) => {
+        if (isHidden) {
+          hiddenIndexes.concat(this.hot.toVisualColumn(physicalIndex));
```
Comment (budnix) [inline src/plugins/hiddenColumns/hiddenColumns.js:172]:
```suggestion
          hiddenIndexes.push(this.hot.toVisualColumn(physicalIndex));
```
The `concat` method doesn't have any effect on the `destinationHideConfig` array. Hence, the hook `beforeUnhideColumns` gets an empty array while showing the columns. Does it seem that there are no tests for those arguments?

File: src/plugins/hiddenRows/hiddenRows.js [hunk 6]
```diff
@@ -154,30 +154,42 @@ class HiddenRows extends BasePlugin {
...
+
+      // Preparing new values for hiding map.
+      arrayEach(physicalRows, (physicalRow) => {
+        hidingMapValues[physicalRow] = false;
+      });
+
+      // Preparing new hiding config.
+      destinationHideConfig = arrayReduce(hidingMapValues, (hiddenIndexes, isHidden, physicalIndex) => {
+        if (isHidden) {
+          hiddenIndexes.concat(this.hot.toVisualRow(physicalIndex));
```
Comment (budnix) [inline src/plugins/hiddenRows/hiddenRows.js:173]:
```suggestion
          hiddenIndexes.push(this.hot.toVisualRow(physicalIndex));
```
The same as [here](https://github.com/handsontable/handsontable/pull/7223/files#r481876518)

File: src/plugins/trimRows/trimRows.js [hunk 7]
```diff
@@ -172,27 +172,37 @@ class TrimRows extends BasePlugin {
...
+    if (isValidConfig && isAnyRowUntrimmed) {
+      // Preparing new values for trimming map.
+      arrayEach(rows, (physicalRow) => {
+        trimmingMapValues[physicalRow] = false;
+      });
+
+      // Preparing new trimming config.
+      destinationTrimConfig = arrayReduce(trimmingMapValues, (trimmedIndexes, isTrimmed, physicalIndex) => {
+        if (isTrimmed) {
+          trimmingMapValues.concat(physicalIndex);
```
Comment (budnix) [inline src/plugins/trimRows/trimRows.js:187]:
```suggestion
          trimmingMapValues.push(physicalIndex);
```
The same as [here](https://github.com/handsontable/handsontable/pull/7223/files#r481876518)

File: src/plugins/hiddenColumns/hiddenColumns.js (hunk 5 above)
Comment (wszymanski) [reply src/plugins/hiddenColumns/hiddenColumns.js:172]: Huh, it looks like it! Thanks for mindfulness. I will add extra test.

**Review by budnix — APPROVED** [review]
🏅

**Pull request merged**